app/services/orders/handle_fees_service.rb

```diff
@@ -16,21 +16,25 @@ def recreate_all_fees!
       # See https://github.com/rails/rails/blob/3-2-stable/activerecord/lib/active_record/locking/pessimistic.rb#L69
       # and https://www.postgresql.org/docs/current/static/sql-select.html#SQL-FOR-UPDATE-SHARE
       order.with_lock do
-        EnterpriseFee.clear_all_adjustments order
+        EnterpriseFee.clear_order_adjustments order
 
-        create_line_item_fees!
+        create_or_update_line_item_fees!
         create_order_fees!
       end
 
       tax_enterprise_fees! unless order.before_payment_state?
       order.update_order!
     end
 
-    def create_line_item_fees!
-      order.line_items.includes(variant: :product).each do |line_item|
-        if provided_by_order_cycle? line_item
-          calculator.create_line_item_adjustments_for line_item
-        end
+    def create_or_update_line_item_fees!
+      order.line_items.includes(:variant).each do |line_item|
+        # No fee associated with the line item so we just create them
+        next create_line_item_fees!(line_item) if line_item.enterprise_fee_adjustments.blank?
```

Comment:
Interesting use of `next` to also execute a command here. It would be a bit clearer to break it up to a couple of lines (inside an if block).

Reply:
I guess this means the next command `create_or_update_line_item_fee` will only ever update, because create has already been handled.
But is this condition needed to skip `update_removed_fees`? I wonder if it is?

Sorry, I don't think I'm following the whole picture very well.

Reply:
`next` works a bit like `return` and can have a return value. So _if_ `line_item.enterprise_fee_adjustments.blank?` is true then the execution of the block is stopped here and the return value of the block is `create_line_item_fees!(line_item)`. The return value is actually ignored because we are in an `each` block. It would matter in a `map` block, for example. So this is confusing. I would also suggest to change this.
```suggestion
        if line_item.enterprise_fee_adjustments.blank?
          create_line_item_fees!(line_item)
          next
        end
```